modules/mtl/arrow-mtl-data/src/main/kotlin/arrow/mtl/AccumT.kt

```diff
@@ -0,0 +1,188 @@
+        }
+      }
+    }
+
+    fun <F, S, A> liftF(MS: Monoid<S>, AF: Applicative<F>, fa: Kind<F, A>): AccumT<S, F, A> = AF.run {
+      AccumT { _ -> fa.map { a -> Tuple2(MS.empty(), a) } }
+    }
+
+    /**
+     * look is an action that fetches all the previously accumulated output.
+     */
+    fun <S, F> look(MS: Monoid<S>, MF: Monad<F>): AccumT<S, F, S> =
+      looks(MS, MF, ::identity)
+
+    /**
+     * looks is an action that retrieves a function of the previously accumulated output.
+     */
+    fun <S, F, A> looks(MS: Monoid<S>, MF: Monad<F>, fs: (S) -> A): AccumT<S, F, A> =
+      AccumT { s: S -> MF.just(MS.empty() toT fs(s)) }
+
+    /**
+     * add w is an action that produces the output w.
+     */
+    fun <S, F> add(MF: Monad<F>, s: S): AccumT<S, F, Unit> =
+      AccumT { _: S -> MF.just(s toT Unit) }
+  }
+
+  /**
+   * Unwrap an accumulation computation.
+   */
+  fun runAccumT(s: S): Kind<F, Tuple2<S, A>> =
+    runAccumT(accumT, s)
+
+  /**
+   * Extract the output from an accumulation computation.
+   */
+  fun execAccumT(MF: Monad<F>, s: S): Kind<F, S> =
+    MF.run {
+      runAccumT(s).map {
+        it.a
+      }
+    }
+
+  /**
+   * Evaluate an accumulation computation with the given initial output history
+   * and return the final value, discarding the final output.
+   */
+  fun evalAccumT(MF: Monad<F>, s: S): Kind<F, A> =
+    MF.run {
+      runAccumT(s).map {
+        it.b
+      }
+    }
+
+  /**
+   * Map both the return value and output of a computation using the given function.
+   */
+  fun <G, B> mapAccumT(
+    fas: (Kind<F, Tuple2<S, A>>) -> Kind<G, Tuple2<S, B>>
+  ): AccumT<S, G, B> =
+    AccumT(
+      AndThen(accumT).andThen(fas)
+    )
+
+  fun <B> map(MF: Functor<F>, fab: (A) -> B): AccumT<S, F, B> = AccumT(
+    AndThen(accumT).andThen {
+      MF.run { it.map { it.a toT fab(it.b) } }
+    }
+  )
+
+  fun <B> flatMap(MS: Monoid<S>, MF: Monad<F>, fa: (A) -> AccumTOf<S, F, B>): AccumT<S, F, B> =
+    AccumT { s: S ->
+      MF.run {
+        runAccumT(s).flatMap { (s1, a) ->
+          fa(a).fix().runAccumT(MS.run { s.combine(s1) }).flatMap { (s2, b) ->
+            MF.just(MS.run { s1.combine(s2) } toT b)
+          }
+        }
+      }
+    }
+
+  fun <B> ap(MS: Monoid<S>, MF: Monad<F>, mf: AccumTOf<S, F, (A) -> B>): AccumT<S, F, B> =
+    (this to mf.fix()).let { (mv, mf) ->
+      AccumT { s: S ->
+        MF.run {
+          runAccumT(mv.accumT, s).flatMap { (s1, v) ->
+            runAccumT(mf.accumT, MS.run { s.combine(s1) }).flatMap { (s2, f) ->
+              MF.just(MS.run { s1.combine(s2) } toT f(v))
+            }
+          }
+        }
+      }
+    }
+
+  /**
+   * Convert an accumulation (append-only) computation into a fully stateful computation.
+   */
+  fun toStateT(MF: Monad<F>, MS: Monoid<S>): StateT<F, S, A> =
+    StateT(MF) { s: S ->
+      MF.run {
+        runAccumT(s).map { (s1, a) ->
+          MS.run { s.combine(s1) } toT a
+        }
+      }
+    }
```

Comment:
i needed to access the outer 's' in the 'andThen' block here. please have a look if i am doing it the right way:

https://github.com/arrow-kt/arrow/pull/1988/commits/4da20729ee3e7fe655e50e7e2e20ccc67ca10a05

Reply:
The idea behind #2005 is that you can do: `StateT(MF, AndThen.id<S>().flatMap { s -> AndThen(runAccumT).andThen { it.map { (s1, a) -> MS.run { s + s1 } toT a } } }))`
And different from before #2005 the inner flatMap is correctly run through a loop and will not cause a stackoverflow

Reply:
got that. i was just curious if what i did  here https://github.com/arrow-kt/arrow/commit/4da20729ee3e7fe655e50e7e2e20ccc67ca10a05 was correct too.

i will use the enhanced `AndThen` from #2005

Reply:
> got that. i was just curious if what i did here 4da2072 was correct too.

Sadly not: Imagine `runAccumT` being `AndThen.Single` with depth `x`, but when you create a new function with `AndThen { ... }` it will set the new depth to `1` but it is actually `x + 1`. My first attempt at solving this problem a while back was to pass this new depth to `AndThen.Single`, but one case was quite hard to solve: `x == maxStackDepth` because you are changing both input and output and keep the state, you to lay one more thing on the stack, so using the usualy `AndThen.Concat` won't work. It's much easier to just define `AndThen.join` instead, also leads to a nice implementation^^